File: app/src/main/java/de/baumann/browser/View/RecordAdapter.java
package de.baumann.browser.View;

import android.content.Context;
import android.support.annotation.NonNull;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ArrayAdapter;
import android.widget.TextView;
import com.github.curioustechizen.ago.RelativeTimeTextView;
import de.baumann.browser.Database.Record;
import de.baumann.browser.Ninja.R;

import java.util.List;

public class RecordAdapter extends ArrayAdapter<Record> {
    private final Context context;
    private final int layoutResId;
    private final List<Record> list;

    public RecordAdapter(Context context, List<Record> list) {
        super(context, R.layout.record_item, list);
        this.context = context;
        this.layoutResId = R.layout.record_item;
        this.list = list;
    }

    private static class Holder {
        TextView title;
        RelativeTimeTextView time;
        TextView url;
    }

    @NonNull
    @Override
    public View getView(int position, View convertView, @NonNull ViewGroup parent) {
        Holder holder;
        View view = convertView;

        if (view == null) {
            view = LayoutInflater.from(context).inflate(layoutResId, parent, false);
            holder = new Holder();
            holder.title = view.findViewById(R.id.record_item_title);
            holder.time = view.findViewById(R.id.record_item_time);
            holder.url = view.findViewById(R.id.record_item_url);
            view.setTag(holder);
        } else {
            holder = (Holder) view.getTag();
        }

        Record record = list.get(position);
        holder.title.setText(record.getTitle());
        holder.time.setReferenceTime(record.getTime());
        holder.url.setText(record.getURL());

        return view;
    }
}

File: app/src/main/java/de/baumann/browser/view/adapter/RecordAdapter.kt
package de.baumann.browser.view.adapter

import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.ImageButton
import android.widget.TextView
import androidx.annotation.LayoutRes
import androidx.recyclerview.widget.RecyclerView
import de.baumann.browser.Ninja.R
import de.baumann.browser.database.Record
import de.baumann.browser.database.RecordType
import java.text.SimpleDateFormat
import java.util.*

class RecordAdapter(
        private val records: MutableList<Record>,
        private val onItemClick: OnItemClick,
        private val onItemLongClick: OnItemLongClick,
) : RecyclerView.Adapter<RecordHolder>() {

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): RecordHolder =
        RecordHolder(parent.inflate(R.layout.list_item, false))

    override fun onBindViewHolder(holder: RecordHolder, position: Int) =
        with (holder) {
            bindRecord(records[position])
            clickAction = { onItemClick(position) }
            longClickAction = { onItemLongClick(position) }
        }

    override fun getItemCount(): Int = records.size

    fun removeAt(position: Int) {
        records.removeAt(position)
        notifyDataSetChanged()
    }

    fun clear() {
        records.clear()
        notifyDataSetChanged()
    }

    fun getItemAt(position: Int): Record = records[position]
}

class RecordHolder(v: View) : RecyclerView.ViewHolder(v), View.OnClickListener, View.OnLongClickListener {
    private val title: TextView = v.findViewById(R.id.record_item_title)
    private val time: TextView = v.findViewById(R.id.record_item_time)
    private val icon: ImageButton = v.findViewById(R.id.ib_icon)

    var clickAction: (() -> Unit)? = null
    var longClickAction: (() -> Unit)? = null

    var record: Record? = null

    init {
        v.setOnClickListener(this)
        v.setOnLongClickListener(this)
    }

    override fun onClick(v: View) = clickAction?.invoke() ?: Unit

    override fun onLongClick(v: View?): Boolean {
        longClickAction?.invoke()
        return true
    }

    fun bindRecord(record: Record) {
        this.record = record
        title.text = record.title
        time.text = SimpleDateFormat("MMM dd", Locale.getDefault()).format(record.time)
        icon.setImageResource(when (record.type) {
            RecordType.Bookmark -> R.drawable.ic_bookmark
            RecordType.History -> R.drawable.icon_earth
        })
    }
}

fun ViewGroup.inflate(@LayoutRes layoutRes: Int, attachToRoot: Boolean = false): View =
    LayoutInflater.from(context).inflate(layoutRes, this, attachToRoot)

typealias OnItemClick = (position: Int) -> Unit
typealias OnItemLongClick = (position: Int) -> Unit
